drivers/ad917x/ad9172.c

```diff
@@ -0,0 +1,391 @@
+		return -EIO;
+	}
+
+	printf("code_grp_sync: %x\n", link_status.code_grp_sync_stat);
+	printf("frame_sync_stat: %x\n", link_status.frame_sync_stat);
+	printf("good_checksum_stat: %x\n",
+	       link_status.good_checksum_stat);
+	printf("init_lane_sync_stat: %x\n",
+	       link_status.init_lane_sync_stat);
+	printf("%d lanes @ %"PRIu64" kBps\n",
+	       st->appJesdConfig.jesd_L, lane_rate_kHz);
+
+	if (st->jesd_dual_link_mode || st->interpolation == 1)
+		dac_mask = AD917X_DAC0 | AD917X_DAC1;
+	else
+		dac_mask = AD917X_DAC0;
+
+	if (st->interpolation > 1) {
+		chan_mask = GENMASK(st->appJesdConfig.jesd_M / 2, 0);
+		ret = ad917x_set_page_idx(ad917x_h, AD917X_DAC_NONE, chan_mask);
+		if (ret != 0)
+			return -EIO;
+		ret = ad917x_set_channel_gain(ad917x_h, 2048); /* GAIN = 1 */
+		if (ret != 0)
+			return -EIO;
+
+		st->nco_main_enable = dac_mask;
+
+		ad917x_nco_enable(ad917x_h, st->nco_main_enable, 0);
+	}
+
+	ret = ad917x_set_page_idx(ad917x_h, dac_mask, AD917X_CH_NONE);
+	if (ret != 0)
+		return -EIO;
+
+	ad917x_register_write(ad917x_h, 0x596, 0x1c); /* SPI turn on TXENx feature */
+
+	return 0;
+}
+
+/**
+ * Delay microseconds, compatible with AD917x API
+ * Performs a blocking or sleep delay for the specified time in microseconds
+ * @param user_data
+ * @param us - time to delay/sleep in microseconds.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+static int32_t ad9172_delay_us(void *user_data, uint32_t us)
+{
+	udelay(us);
+
+	return 0;
+}
+
+/**
+ * Spi write and read compatible with
+ * @param user_data
+ * @param wbuf Pointer to array with the data to be sent on the SPI
+ * @param rbuf Pointer to array where the data to which the SPI will be written
+ * @param len The size in bytes allocated for each of the indata and outdata arrays.
+ * @return 0 for success, any non-zero value indicates an error
+ */
+static int32_t ad9172_spi_xfer(void *user_data, uint8_t *wbuf,
+			       uint8_t *rbuf, int32_t len)
+{
+	struct spi_desc *spi = user_data;
+	uint8_t * buffer = (uint8_t *) malloc(len);
+	buffer[0] = wbuf[0];
+	buffer[1] = wbuf[1];
+	buffer[2] = wbuf[2];
+	int32_t ret = spi_write_and_read(spi, buffer, len);
+	if (ret < 0) {
+		printf("Read Error %"PRId32, ret);
+	} else {
+		memcpy(rbuf, buffer, len);
+	}
+	free(buffer);
+
+	return ret;
+}
+
+/**
+ * Initialize the device.
+ * @param device - The device structure.
+ * @param init_param - The structure that contains the device initial
+ * 		       parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t ad9172_init(ad9172_dev **device,
+		    ad9172_init_param * init_param)
+{
+	int32_t ret;
+	ad9172_dev *dev;
+
+	dev = (ad9172_dev *)calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	struct ad9172_state *st = (struct ad9172_state *)calloc(1, sizeof(*st));
+	if (!st)
+		goto error_1;
+
+	dev->st = st;
+
+	/* SPI */
+	ret = spi_init(&dev->spi_desc, init_param->spi_init);
+
+	if (ret < 0)
+		return ret;
```

Comment:
if this fails you still have to free memory should be goto error_1, but then you will have to change the return there. 

ret = spi_init(&dev->spi_desc, init_param->spi_init);
if (ret < 0) {
                ret = -EINVAL;
		goto error_1;
}